pyquil/api/compiler.py

```diff
@@ -0,0 +1,148 @@
+from ._base_connection import TYPE_MULTISHOT, get_job_id, get_session, \
+    wait_for_job, post_json, get_json
+
+
+class CompilerConnection(object):
+    """
+    Represents a connection to the Quil compiler.
+    """
+
+    def __init__(self, sync_endpoint='https://api.rigetti.com',
+                 async_endpoint='https://job.rigetti.com/beta', api_key=None,
+                 user_id=None, use_queue=False, ping_time=0.1, status_time=2,
+                 default_isa=None):
+        """
+        Constructor for CompilerConnection. Sets up any necessary security.
+
+        :param sync_endpoint: The endpoint of the server for running small jobs
+        :param async_endpoint: The endpoint of the server for running large jobs
+        :param api_key: The key to the Forest API Gateway (default behavior is
+                        to read from config file)
+        :param user_id: Your userid for Forest (default behavior is to read from
+                        config file)
+        :param bool use_queue: Disabling this parameter may improve performance
+                               for small, quick programs. To support larger
+                               programs, set it to True. (default: False)
+                               NOTE: *_async methods will always use the queue.
+                               See https://go.rigetti.com/connections for more
+                               information.
+        :param int ping_time: Time in seconds for how long to wait between
+                              polling the server for updated status information
+                              on a job. Note that this parameter doesn't matter
+                              if use_queue is False.
+        :param int status_time: Time in seconds for how long to wait between
+                                printing status information. To disable printing
+                                of status entirely then set status_time to
+                                False. Note that this parameter doesn't matter
+                                if use_queue is False.
+        :param ISA default_isa: A default ISA object to target when one is not
+                            expressly provided to the compile method.
+        """
+        self.async_endpoint = async_endpoint
+        self.sync_endpoint = sync_endpoint
+        self.session = get_session(api_key, user_id)
+
+        self.use_queue = use_queue
+        self.ping_time = ping_time
+        self.status_time = status_time
+
+        self.default_isa = default_isa
+
+    def compile(self, quil_program, isa=None):
+        """
+        Sends a Quil program to the Forest compiler and returns the resulting
+        compiled Program.
+
+        :param Program quil_program: Quil program to be compiled.
+        :param ISA isa: ISA to target.
+        :returns: The compiled Program object.
+        :rtype: Program
+        """
+        if not isa:
+            isa = self.default_isa
```

Comment:
What happens if default_isa is also None?

Reply:
The remote quilc will return an error. We could preempt this with a raise right away.

Reply:
Ah, indeed, we do raise an error in the payload formation function.